lib/Dialect/FIRRTL/Transforms/IMDCE.cpp

```diff
@@ -0,0 +1,412 @@
+
+  /// This keeps track of users the instance results that correspond to output
+  /// ports.
+  DenseMap<BlockArgument, llvm::TinyPtrVector<Value>>
+      resultPortToInstanceResultMapping;
+  InstanceGraph *instanceGraph;
+
+  /// A worklist of values whose liveness recently changed, indicating the
+  /// users need to be reprocessed.
+  SmallVector<Value, 64> worklist;
+  llvm::DenseSet<Value> liveSet;
+};
+} // namespace
+
+void IMDCEPass::markWireOrRegOrNode(Operation *op) {
+  assert(isWireOrRegOrNode(op) && "only a wire, a reg or a node is expected");
+  auto resultValue = op->getResult(0);
+
+  if (!isDeletableWireOrRegOrNode(op))
+    markAlive(resultValue);
+}
+
+void IMDCEPass::markMemOp(MemOp mem) {
+  for (auto result : mem.getResults())
+    markAlive(result);
+}
+
+void IMDCEPass::markUnknownSideEffectOp(Operation *op) {
+  // For operations with side effects, pessimistically mark results and
+  // operands as alive.
+  for (auto result : op->getResults())
+    markAlive(result);
+  for (auto operand : op->getOperands())
+    markAlive(operand);
+}
+
+void IMDCEPass::visitUser(Operation *op) {
+  LLVM_DEBUG(llvm::dbgs() << "Visit: " << *op << "\n");
+  if (auto connectOp = dyn_cast<FConnectLike>(op))
+    return visitConnect(connectOp);
+  if (isa<SubfieldOp, SubindexOp, SubaccessOp>(op))
+    return visitSubelement(op);
+}
+
+void IMDCEPass::markInstanceOp(InstanceOp instance) {
+  // Get the module being reference.
+  Operation *op = instanceGraph->getReferencedModule(instance);
+
+  // If this is an extmodule, just remember that any inputs and inouts are
+  // alive.
+  if (!isa<FModuleOp>(op)) {
+    auto module = dyn_cast<FModuleLike>(op);
+    for (size_t resultNo = 0, e = instance.getNumResults(); resultNo != e;
+         ++resultNo) {
+      auto portVal = instance.getResult(resultNo);
+      // If this is an output to the extmodule, we can ignore it.
+      if (module.getPortDirection(resultNo) == Direction::Out)
+        continue;
+
+      // Otherwise this is an inuput from it or an inout, mark it as alive.
+      markAlive(portVal);
+    }
+    return;
+  }
+
+  // Otherwise this is a defined module.
+  auto fModule = cast<FModuleOp>(op);
+  markBlockExecutable(fModule.getBody());
+
+  // Ok, it is a normal internal module reference so populate
+  // resultPortToInstanceResultMapping.
+  for (size_t resultNo = 0, e = instance.getNumResults(); resultNo != e;
+       ++resultNo) {
+    auto instancePortVal = instance.getResult(resultNo);
+
+    // Otherwise we have a result from the instance.  We need to forward results
+    // from the body to this instance result's SSA value, so remember it.
+    BlockArgument modulePortVal = fModule.getArgument(resultNo);
+
+    // Mark don't touch ports as alive.
+    if (hasDontTouch(modulePortVal)) {
+      markAlive(modulePortVal);
+      markAlive(instancePortVal);
+    }
+
+    resultPortToInstanceResultMapping[modulePortVal].push_back(instancePortVal);
```

Comment:
I'm  not totally sure I wrapped my head around all of the logic here.

It would make more sense to me if we check for DontTouch on block arguments inside `markBlockExecutable`.  

Since you [check](https://github.com/llvm/circt/pull/3324/files#diff-59ce1ccf622a31c1b31f9342839822bddcf8c2687ad403a483146d829ecb40b8R324-R326) for DontTouch in RewriteModuleSignature, do we have to propagate this DontTouch to the instance at all here?

Reply:
You are right. I moved dontTouch marking to `makkBlockExecutable`. 

> Since you [check](https://github.com/llvm/circt/pull/3324/files#diff-59ce1ccf622a31c1b31f9342839822bddcf8c2687ad403a483146d829ecb40b8R324-R326) for DontTouch in RewriteModuleSignature, do we have to propagate this DontTouch (markAlive(instancePortVal)) to the instance at all here?

Yeah, I agree we don't have to mark instancePortval as alive here.